File: app/src/main/java/com/yasmin/solace_application/ui/ChatbotActivity.java
package com.yasmin.solace_application.ui;

import android.os.Bundle;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.JsonObjectRequest;
import com.android.volley.toolbox.Volley;
import com.yasmin.solace_application.R;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

public class ChatbotActivity extends AppCompatActivity {


    private RecyclerView chatsRV;
    private ImageButton sendMsgIB;
    private EditText userMsgEdt;
    private final String USER_KEY = "user";
    private final String BOT_KEY = "bot";


    private RequestQueue RequestQueue;

    private ArrayList<MsgModal> chatModalArrayList;
    private ChatbotRVAdapter chatsRVAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_chatbot);

        chatsRV = findViewById(R.id.idRVChats);
        sendMsgIB = findViewById(R.id.idIBSend);
        userMsgEdt = findViewById(R.id.idEdtMessage);

        RequestQueue = Volley.newRequestQueue(ChatbotActivity.this);
        RequestQueue.getCache().clear();

        chatModalArrayList = new ArrayList<>();

        sendMsgIB.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                if (userMsgEdt.getText().toString().isEmpty()) {
                    Toast.makeText(ChatbotActivity.this, "Please enter your message..", Toast.LENGTH_SHORT).show();
                    return;
                }


                sendMessage(userMsgEdt.getText().toString());

                userMsgEdt.setText("");
            }
        });

        chatsRVAdapter = new ChatbotRVAdapter(chatModalArrayList, this);


        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(ChatbotActivity.this, RecyclerView.VERTICAL, false);


        chatsRV.setLayoutManager(linearLayoutManager);


        chatsRV.setAdapter(chatsRVAdapter);
    }

    private void sendMessage(String userMsg) {

        chatModalArrayList.add(new MsgModal(userMsg, USER_KEY));
        chatsRVAdapter.notifyDataSetChanged();


        String url = "http://api.brainshop.ai/get?bid=169801&key=gOpx38XQRwRCt8F4&uid=[uid]&msg=" + userMsg;

        RequestQueue queue = Volley.newRequestQueue(ChatbotActivity.this);

        JsonObjectRequest jsonObjectRequest = new JsonObjectRequest(Request.Method.GET, url, null, new Response.Listener<JSONObject>() {
            @Override
            public void onResponse(JSONObject response) {
                try {

                    String botResponse = response.getString("cnt");
                    chatModalArrayList.add(new MsgModal(botResponse, BOT_KEY));


                    chatsRVAdapter.notifyDataSetChanged();
                } catch (JSONException e) {
                    e.printStackTrace();


                    chatModalArrayList.add(new MsgModal("No response", BOT_KEY));
                    chatsRVAdapter.notifyDataSetChanged();
                }
            }
        }, new Response.ErrorListener() {
            @Override
            public void onErrorResponse(VolleyError error) {
                // error handling.
                chatModalArrayList.add(new MsgModal("Sorry no response found", BOT_KEY));
                Toast.makeText(ChatbotActivity.this, "No response from the bot..", Toast.LENGTH_SHORT).show();
            }
        });


        queue.add(jsonObjectRequest);
    }
}

File: app/src/main/java/com/yasmin/solace_application/ui/ChatbotActivity.kt
package com.yasmin.solace_application.ui

import android.annotation.SuppressLint
import android.os.Bundle
import android.util.Log
import android.widget.EditText
import android.widget.ImageButton
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.RecyclerView
import com.google.api.gax.core.FixedCredentialsProvider
import com.google.auth.oauth2.GoogleCredentials
import com.google.auth.oauth2.ServiceAccountCredentials
import com.google.cloud.dialogflow.v2.*
import com.google.cloud.dialogflow.v2beta1.SessionName
import com.yasmin.solace_application.R
import com.yasmin.solace_application.models.Message
import kotlinx.coroutines.Dispatchers.Default
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.util.ArrayList
import java.util.UUID

class ChatbotActivity : AppCompatActivity() {
    private var messageList: ArrayList<Message> =ArrayList()
    lateinit var sessionsClient: SessionsClient
    lateinit var sessionName: SessionName
    private val uuid = UUID.randomUUID().toString()
    val TAG = "mainactivity"
    lateinit var chatAdapter: ChatAdapter
    lateinit var btnSend:ImageButton
    lateinit var chatView:RecyclerView
    lateinit var editMessage:EditText

    @SuppressLint("NotifyDataSetChanged")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_chatbot)

        btnSend=findViewById(R.id.btnSend)
        chatView=findViewById(R.id.chatView)
        editMessage=findViewById(R.id.editMessage)
        chatAdapter = ChatAdapter(this, messageList)
        chatView.adapter = chatAdapter

        btnSend.setOnClickListener {
            val message: String = editMessage.text.toString()
            if (message.isNotEmpty()) {
                addMessageToList(message, false)
                sendMessageToBot(message)
            } else {
                Toast.makeText(this@ChatbotActivity, "Please enter text!", Toast.LENGTH_SHORT).show()
            }
        }

        setUpBot()
    }

    @SuppressLint("NotifyDataSetChanged")
    private fun addMessageToList(message: String, isReceived: Boolean) {
        messageList.add(Message(message, isReceived))
        editMessage.setText("")
        chatAdapter.notifyDataSetChanged()
        chatView.layoutManager?.scrollToPosition(messageList.size - 1)
    }

    private fun setUpBot() {
        try {
            val stream = this.resources.openRawResource(R.raw.credential)
            val credentials: GoogleCredentials = GoogleCredentials.fromStream(stream)
                .createScoped("https://www.googleapis.com/auth/cloud-platform")
            val projectId: String = (credentials as ServiceAccountCredentials).projectId
            val settingsBuilder: SessionsSettings.Builder = SessionsSettings.newBuilder()
            val sessionsSettings: SessionsSettings = settingsBuilder.setCredentialsProvider(
                FixedCredentialsProvider.create(credentials)
            ).build()
            sessionsClient = SessionsClient.create(sessionsSettings)
            sessionName = SessionName.of(projectId, uuid)
            Log.d(TAG, "projectId : $projectId")
        } catch (e: Exception) {
            Log.e(TAG, "setUpBot: " + e.message)
        }
    }

    private fun sendMessageToBot(message: String) {
        val input = QueryInput.newBuilder()
            .setText(TextInput.newBuilder().setText(message).setLanguageCode("en-US")).build()
        GlobalScope.launch {
            sendMessageInBg(input)
        }
    }

    private suspend fun sendMessageInBg(
        queryInput: QueryInput
    ) {
        withContext(Default) {
            try {
                val detectIntentRequest = DetectIntentRequest.newBuilder()
                    .setSession(sessionName.toString())
                    .setQueryInput(queryInput)
                    .build()
                val result = sessionsClient.detectIntent(detectIntentRequest)
                if (result != null) {
                    runOnUiThread {
                        updateUI(result)
                    }
                }
            } catch (e: java.lang.Exception) {
                Log.d(TAG, "doInBackground: " + e.message)
                e.printStackTrace()
            }
        }
    }

    private fun updateUI(response: DetectIntentResponse) {
        val botReply: String = response.queryResult.fulfillmentText
        if (botReply.isNotEmpty()) {
            addMessageToList(botReply, true)
        } else {
            Toast.makeText(this, "something went wrong", Toast.LENGTH_SHORT).show()
        }
    }
}